### PR TITLE
UCP/EP: Add eager multi-fragment overhead

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/ucp/core/ucp_ep.c [hunk 1]
```diff
@@ -1749,7 +1751,14 @@ static void ucp_ep_config_calc_params(ucp_worker_h worker,
             }
         }
 
-        params->bw += ucp_tl_iface_bandwidth(context, &iface_attr->bandwidth);
+        bw = ucp_tl_iface_bandwidth(context, &iface_attr->bandwidth);
+
+        if (eager) {
+            params->bw += 1.0 / ((1.0 / bw) + (iface_attr->overhead /
+                                               iface_attr->cap.am.max_bcopy));
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:1758]: is it possible max_bcopy would be 0?

File: src/ucp/core/ucp_ep.c [hunk 2]
```diff
@@ -1749,7 +1751,14 @@ static void ucp_ep_config_calc_params(ucp_worker_h worker,
             }
         }
 
-        params->bw += ucp_tl_iface_bandwidth(context, &iface_attr->bandwidth);
+        bw = ucp_tl_iface_bandwidth(context, &iface_attr->bandwidth);
+
+        if (eager) {
+            params->bw += 1.0 / ((1.0 / bw) + (iface_attr->overhead /
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:1757]: pls add comment:
`/* eager protocol has overhead for each fragment */`

File: src/ucp/core/ucp_ep.c [hunk 3]
```diff
@@ -1749,7 +1751,14 @@ static void ucp_ep_config_calc_params(ucp_worker_h worker,
             }
         }
 
-        params->bw += ucp_tl_iface_bandwidth(context, &iface_attr->bandwidth);
+        bw = ucp_tl_iface_bandwidth(context, &iface_attr->bandwidth);
+
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:1755]: can remove space line

File: test/gtest/ucp/test_ucp_peer_failure.cc [hunk 4]
```diff
@@ -360,13 +360,21 @@ void test_ucp_peer_failure::do_test(size_t msg_size, int pre_msg_count,
         EXPECT_NE(UCS_OK, m_err_status);
 
         if (UCS_PTR_IS_PTR(sreq)) {
-            /* The request may either succeed or fail, even though the data is
+            ucs_status_t status;
+            /* TODO update comment
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_peer_failure.cc:364]: 1. why needed? if m_err_count!=0, i would expect the request to also complete
2. update comment
3. can we use request_wait()?

Comment (yosefe) [reply test/gtest/ucp/test_ucp_peer_failure.cc:364]: discussed offline:
1. use request_wait
2. update comment: "If rendezvous protocol is used, the m_err_count is increased on the receiver side, so the send request may not complete immediately"

File: src/ucp/core/ucp_ep.c (hunk 2 above)
Comment (yosefe) [reply src/ucp/core/ucp_ep.c:1757]: use
overhead = send_post_overhead + send_pre_overhead (from estimate_perf)

File: src/ucp/core/ucp_ep.c [hunk 5]
```diff
@@ -1749,7 +1752,24 @@ static void ucp_ep_config_calc_params(ucp_worker_h worker,
...
+        if (eager && iface_attr->cap.am.max_bcopy) {
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:1757]: && (iface_attr->cap.am.max_bcopy > 0)

File: src/ucp/core/ucp_ep.c [hunk 6]
```diff
@@ -1749,7 +1752,24 @@ static void ucp_ep_config_calc_params(ucp_worker_h worker,
...
+            /* Eager protocol has overhead for each fragment */
+            perf_attr.field_mask = UCT_PERF_ATTR_FIELD_OPERATION |
+                                   UCT_PERF_ATTR_FIELD_SEND_PRE_OVERHEAD |
+                                   UCT_PERF_ATTR_FIELD_SEND_POST_OVERHEAD;
+            perf_attr.operation  = UCT_EP_OP_AM_ZCOPY;
+
+            uct_iface_estimate_perf(ucp_worker_iface(worker, rsc_index)->iface,
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:1764]: check return value

Comment (Artemy-Mellanox) [reply src/ucp/core/ucp_ep.c:1764]: assert or fail ucp_ep_config_init?

File: src/ucp/core/ucp_ep.c [hunk 7]
```diff
@@ -1754,15 +1756,16 @@ static void ucp_ep_config_calc_params(ucp_worker_h worker,
...
             perf_attr.field_mask = UCT_PERF_ATTR_FIELD_OPERATION |
                                    UCT_PERF_ATTR_FIELD_SEND_PRE_OVERHEAD |
                                    UCT_PERF_ATTR_FIELD_SEND_POST_OVERHEAD;
             perf_attr.operation  = UCT_EP_OP_AM_ZCOPY;
 
-            uct_iface_estimate_perf(ucp_worker_iface(worker, rsc_index)->iface,
-                                    &perf_attr);
+            wiface = ucp_worker_iface(worker, rsc_index);
+            status = uct_iface_estimate_perf(wiface->iface, &perf_attr);
+            ucs_assert(status == UCS_OK);
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:1768]: fail config_init

File: src/ucp/core/ucp_ep.c [hunk 8]
```diff
@@ -1812,17 +1826,23 @@ static size_t ucp_ep_config_calc_rndv_thresh(ucp_worker_t *worker,
                                 (2 * eager_zcopy.overhead) + rndv.overhead) -
                 eager_zcopy.reg_overhead - eager_zcopy.overhead;
 
-    denumerator = eager_zcopy.reg_growth +
+    denominator = eager_zcopy.reg_growth +
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:1829]: yeah..

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@Artemy-Mellanox can you pls squash?

**Review by yosefe — APPROVED** [review]

**Pull request merged**